compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -839,7 +839,8 @@ object PatternMatcher {
               default
           }
         case ResultPlan(tree) =>
-          Return(tree, ref(resultLabel))
+          if (tree.symbol eq defn.throwMethod) tree // Namely MatchError
```

Comment:
Don't you want to be more general and test something like `tree.tpe <:< NothingTpe` instead?

Reply:
Sounds good